node/src/store.rs

```diff
@@ -1,14 +1,296 @@
+        self.get_by_hash(hash)
+    }
+
+    async fn get_by_height(&self, height: u64) -> Result<ExtendedHeader, StoreError> {
+        self.get_by_height(height)
+    }
+
+    async fn height(&self) -> u64 {
+        self.get_head_height()
+    }
+
+    async fn has(&self, hash: &Hash) -> bool {
+        self.exists_by_hash(hash)
+    }
+
+    async fn has_at(&self, height: u64) -> bool {
+        self.exists_by_height(height)
+    }
+
+    async fn append_single(&mut self, header: ExtendedHeader) -> Result<()> {
+        self.append_continuous(header)
+    }
+}
+
 #[derive(Debug)]
-pub struct Store {}
+pub struct InMemoryStore {
+    headers: DashMap<Hash, ExtendedHeader>,
+    height_to_hash: DashMap<u64, Hash>,
+    head_height: AtomicU64,
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum StoreError {
+    // TODO: error consolidation again
+    #[error("Hash {0} already exists in store")]
+    HashExists(Hash),
+    #[error("Height {0} already exists in store")]
+    HeightExists(u64),
 
-impl Store {
+    #[error("Continuous append impossible")]
+    NonContinuousAppend,
+    #[error("Failed to apply header {0}")]
+    ContinuousAppendFailedAt(usize),
+
+    #[error("Header not found in store")]
+    NotFound,
+
+    #[error("Store in inconsistent state, lost head")]
+    LostStoreHead,
+    #[error("Store in inconsistent state; height->hash mapping exists, {0} missing")]
+    LostHash(Hash),
+}
+
+impl InMemoryStore {
     pub fn new() -> Self {
-        Store {}
+        InMemoryStore {
+            headers: DashMap::new(),
+            height_to_hash: DashMap::new(),
+            head_height: AtomicU64::new(0),
+        }
+    }
+
+    pub fn with_genesis(genesis: ExtendedHeader) -> Self {
+        let genesis_hash = genesis.hash();
+        let genesis_height = genesis.height().value();
+
+        InMemoryStore {
+            headers: DashMap::from_iter([(genesis_hash, genesis)]),
+            height_to_hash: DashMap::from_iter([(genesis_height, genesis_hash)]),
+            head_height: AtomicU64::new(genesis_height),
+        }
+    }
+
+    pub fn get_head_height(&self) -> u64 {
+        self.head_height.load(Ordering::Acquire)
+    }
+
+    #[instrument(err)]
+    pub fn append_continuous(&self, header: ExtendedHeader) -> Result<(), StoreError> {
+        let hash = header.hash();
+        let height = header.height();
+
+        // lock both maps to ensure consistency
+        // this shouldn't deadlock as long as we don't hold references across awaits if any
+        // https://github.com/xacrimon/dashmap/issues/233
+        let hash_entry = self.headers.entry(hash);
+        let height_entry = self.height_to_hash.entry(height.into());
+
+        if matches!(hash_entry, Entry::Occupied(_)) {
+            return Err(StoreError::HashExists(hash));
+        }
+        if matches!(height_entry, Entry::Occupied(_)) {
+            return Err(StoreError::HeightExists(height.into()));
+        }
```

Comment:
We need to check about this. Some amount of the most current blocks can have the same height but different hashes.

Should the Syncer decide which chain wins?

Reply:
After talking with the Celestia team, they told me that it is impossible to have a block with same height but different hashes. So we can say that this if-statement is valid.